File: Sources/HerdrChatUI/Views/Components.swift
import SwiftUI
import Foundation
import HerdrKit

// MARK: - Signature: presence ring avatar

/// The app's one signature element: a workspace avatar wrapped in a thin ring
/// that *breathes* in emerald while the agent works and holds solid orange when
/// it's waiting on you — a terminal cursor's blink, translated to iOS. Idle
/// workspaces show no ring at all, so attention goes exactly where it should.
struct PresenceRingAvatar: View {
    let title: String
    let key: String
    let status: AgentStatus
    var size: CGFloat = 52

    private var ring: Color? {
        switch status {
        case .working: return Theme.tint
        case .blocked: return Theme.attention
        case .done: return Theme.tint.opacity(0.45)
        case .idle, .unknown: return nil
        }
    }

    var body: some View {
        ZStack {
            if let ring {
                if status == .working {
                    TimelineView(.animation) { context in
                        let t = context.date.timeIntervalSinceReferenceDate
                        Circle()
                            .strokeBorder(ring, lineWidth: 2)
                            .opacity(0.35 + 0.65 * (0.5 + 0.5 * sin(t * 2.4)))
                    }
                } else {
                    Circle().strokeBorder(ring, lineWidth: 2)
                }
            }
            Circle()
                .fill(Avatar.color(for: key))
                .frame(width: size, height: size)
                .overlay(
                    Text(initials(of: title))
                        .font(.system(size: size * 0.36, weight: .semibold, design: .rounded))
                        .foregroundStyle(.white)
                )
        }
        .frame(width: size + 8, height: size + 8)
    }
}

func initials(of title: String) -> String {
    let letters = title.split(separator: " ").compactMap(\.first).prefix(2)
    return letters.isEmpty ? "?" : String(letters).uppercased()
}

// MARK: - Typing indicator

/// Three softly pulsing dots, driven by the display clock.
struct TypingDots: View {
    var color: Color = .secondary
    var size: CGFloat = 5

    var body: some View {
        TimelineView(.animation) { context in
            let t = context.date.timeIntervalSinceReferenceDate
            HStack(spacing: 3) {
                ForEach(0..<3, id: \.self) { i in
                    let phase = sin(t * 5 - Double(i) * 0.9)
                    Circle()
                        .fill(color)
                        .frame(width: size, height: size)
                        .opacity(0.3 + 0.7 * (phase * 0.5 + 0.5))
                }
            }
        }
        .frame(height: size)
    }
}

// MARK: - Message bubble

/// A native-feeling chat bubble: 18pt continuous corners, no shadow, system
/// surfaces. Following the iMessage convention, only the LAST bubble of a run
/// gets the small "tail" corner; bubbles inside a run stay fully rounded.
struct MessageBubble: View {
    let message: ChatMessage
    var isLastInGroup: Bool = true

    private var isOutgoing: Bool { message.role == .user }

    var body: some View {
        HStack {
            if isOutgoing { Spacer(minLength: 48) }
            VStack(alignment: .leading, spacing: 4) {
                ForEach(Array(message.segments.enumerated()), id: \.offset) { _, segment in
                    segmentView(segment)
                }
                if isLastInGroup, let time = formatTime(message.timestamp) {
                    Text(time)
                        .font(.caption2)
                        .foregroundStyle(isOutgoing ? Color.white.opacity(0.7) : Color.secondary.opacity(0.8))
                        .frame(maxWidth: .infinity, alignment: .trailing)
                }
            }
            .padding(.horizontal, 12)
            .padding(.vertical, 7)
            .background(isOutgoing ? Theme.tint : Theme.bubbleIncoming)
            .clipShape(bubbleShape)
            .frame(maxWidth: 300, alignment: isOutgoing ? .trailing : .leading)
            if !isOutgoing { Spacer(minLength: 48) }
        }
    }

    private var bubbleShape: UnevenRoundedRectangle {
        let r: CGFloat = 18
        let tail: CGFloat = 5
        if isOutgoing {
            return UnevenRoundedRectangle(
                topLeadingRadius: r,
                bottomLeadingRadius: r,
                bottomTrailingRadius: isLastInGroup ? tail : r,
                topTrailingRadius: r,
                style: .continuous
            )
        } else {
            return UnevenRoundedRectangle(
                topLeadingRadius: r,
                bottomLeadingRadius: isLastInGroup ? tail : r,
                bottomTrailingRadius: r,
                topTrailingRadius: r,
                style: .continuous
            )
        }
    }

    @ViewBuilder
    private func segmentView(_ segment: MessageSegment) -> some View {
        switch segment {
        case .text(let text):
            MarkdownText(markdown: text, foreground: isOutgoing ? .white : .primary, onTint: isOutgoing)
                .font(.body)
                .textSelection(.enabled)
        case .thinking:
            TerminalChip(glyph: "…", label: "thought", glyphColor: .secondary)
        case .toolUse(let name, let input):
            TerminalChip(glyph: "❯", label: input.map { "\(name) \($0)" } ?? name, glyphColor: Theme.tint, expandable: input != nil)
        case .toolResult:
            TerminalChip(glyph: "✓", label: "tool result", glyphColor: .secondary)
        }
    }
}

/// Tool activity as a terminal token: monospaced, quiet, unmistakably CLI. When
/// `expandable`, tapping toggles between a one-line preview and the full text, so
/// a long tool call can be read in place instead of staying truncated.
private struct TerminalChip: View {
    let glyph: String
    let label: String
    let glyphColor: Color
    var expandable: Bool = false
    @State private var expanded = false

    var body: some View {
        HStack(alignment: .top, spacing: 5) {
            Text(glyph)
                .font(.system(.caption2, design: .monospaced).weight(.bold))
                .foregroundStyle(glyphColor)
            Text(label)
                .font(.system(.caption2, design: .monospaced))
                .foregroundStyle(.secondary)
                .lineLimit(expandable && expanded ? nil : 1)
                .textSelection(.enabled)
            if expandable {
                Image(systemName: expanded ? "chevron.up" : "chevron.down")
                    .font(.system(size: 8, weight: .bold))
                    .foregroundStyle(.tertiary)
            }
        }
        .padding(.horizontal, 7)
        .padding(.vertical, 3)
        .background(Theme.fillSubtle)
        .clipShape(RoundedRectangle(cornerRadius: 6, style: .continuous))
        .contentShape(Rectangle())
        .onTapGesture {
            guard expandable else { return }
            withAnimation(.easeInOut(duration: 0.15)) { expanded.toggle() }
        }
    }
}

// MARK: - Live preview (agent's in-progress answer)

/// A dim trailing bubble showing the agent's answer as it's being written
/// (scraped from the pane's visible screen). Superseded by the real bubble when
/// the turn settles. Only shown when clean prose was extracted — otherwise the
/// waiting bar shows instead.
struct LivePreviewBubble: View {
    let text: String

    var body: some View {
        HStack {
            VStack(alignment: .leading, spacing: 6) {
                HStack(spacing: 6) {
                    TypingDots(color: Theme.tint, size: 4)
                    Text("live")
                        .font(.system(.caption2, design: .rounded).weight(.semibold))
                        .foregroundStyle(Theme.tint)
                }
                Text(text)
                    .font(.callout)
                    .foregroundStyle(.secondary)
                    .frame(maxWidth: .infinity, alignment: .leading)
            }
            .padding(.horizontal, 12)
            .padding(.vertical, 8)
            .background(Theme.bubbleIncoming.opacity(0.6))
            .clipShape(
                UnevenRoundedRectangle(topLeadingRadius: 5, bottomLeadingRadius: 18, bottomTrailingRadius: 18, topTrailingRadius: 18, style: .continuous)
            )
            .frame(maxWidth: 300, alignment: .leading)
            Spacer(minLength: 48)
        }
    }
}

// MARK: - Waiting indicator

/// A slim, indeterminate "waiting for the reply" bar shown while the agent
/// works. Transcripts are turn-granular, so there's no finer token stream to
/// surface — a quiet sweeping bar reads as progress without pretending to show
/// content it doesn't have (it replaces the old terminal-tail bubble).
struct WaitingBar: View {
    var height: CGFloat = 3

    var body: some View {
        GeometryReader { geo in
            let w = geo.size.width
            let segment = max(48, w * 0.28)
            ZStack(alignment: .leading) {
                Capsule().fill(Theme.tint.opacity(0.14))
                TimelineView(.animation) { context in
                    let t = context.date.timeIntervalSinceReferenceDate
                    // Ease-in-out sweep back and forth across the track.
                    let phase = 0.5 - 0.5 * cos(t * 1.5)
                    let x = (w - segment) * phase
                    Capsule()
                        .fill(Theme.tint)
                        .frame(width: segment)
                        .offset(x: x)
                }
            }
        }
        .frame(height: height)
        .clipShape(Capsule())
        .accessibilityLabel("Waiting for reply")
    }
}

// MARK: - Blocked quick replies

/// Shown when an agent is waiting for input. When the pane's choice menu could
/// be parsed, it shows the actual question and one full-width button per option
/// labelled with its real text (e.g. "2. Yes, and don't ask again"), so you can
/// see what you're picking. Otherwise it falls back to generic QuickType chips.
struct BlockedReplyBar: View {
    /// Parsed choice menu from the blocked pane, if any.
    var prompt: BlockedPrompt?
    let onKeys: ([String]) -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            if let prompt, !prompt.isEmpty {
                labelledOptions(prompt)
            } else {
                genericChips
            }
        }
        .padding(.horizontal, 14)
        .padding(.vertical, 10)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(.bar)
    }

    @ViewBuilder
    private func labelledOptions(_ prompt: BlockedPrompt) -> some View {
        Label(prompt.question ?? "Agent is waiting", systemImage: "exclamationmark.bubble.fill")
            .font(.footnote.weight(.medium))
            .foregroundStyle(Theme.attention)
            .fixedSize(horizontal: false, vertical: true)
        ForEach(prompt.options) { option in
            Button {
                onKeys(option.keys)
            } label: {
                HStack(alignment: .firstTextBaseline, spacing: 8) {
                    Text("\(option.number)")
                        .font(.footnote.weight(.bold).monospacedDigit())
                        .foregroundStyle(Theme.attention)
                        .frame(minWidth: 16, alignment: .trailing)
                    Text(option.label)
                        .font(.subheadline)
                        .foregroundStyle(.primary)
                        .multilineTextAlignment(.leading)
                        .fixedSize(horizontal: false, vertical: true)
                    Spacer(minLength: 0)
                }
                .padding(.horizontal, 12)
                .padding(.vertical, 9)
                .frame(maxWidth: .infinity, alignment: .leading)
                .background(
                    RoundedRectangle(cornerRadius: 12, style: .continuous)
                        .fill(Theme.attention.opacity(0.12))
                )
            }
            .buttonStyle(PressableStyle())
        }
    }

    private var genericChips: some View {
        VStack(alignment: .leading, spacing: 7) {
            Label("Agent is waiting", systemImage: "exclamationmark.bubble.fill")
                .font(.footnote.weight(.medium))
                .foregroundStyle(Theme.attention)
            ScrollView(.horizontal, showsIndicators: false) {
                HStack(spacing: 8) {
                    chip("Confirm", icon: "return", keys: ["Enter"])
                    chip("1", keys: ["1", "Enter"])
                    chip("2", keys: ["2", "Enter"])
                    chip("Esc", icon: "escape", keys: ["Escape"])
                }
            }
        }
    }

    @ViewBuilder
    private func chip(_ title: String, icon: String? = nil, keys: [String]) -> some View {
        Button {
            onKeys(keys)
        } label: {
            if let icon {
                Label(title, systemImage: icon)
            } else {
                Text(title)
            }
        }
        .buttonStyle(.bordered)
        .buttonBorderShape(.capsule)
        .controlSize(.small)
        .tint(Theme.attention)
    }
}

File: android/app/src/main/java/dev/herdr/herdrchat/ui/components/Components.kt
package dev.herdr.herdrchat.ui.components

import androidx.compose.animation.core.FastOutSlowInEasing
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Build
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.Psychology
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import dev.herdr.herdrchat.core.model.AgentStatus
import dev.herdr.herdrchat.core.transcript.ChatMessage
import dev.herdr.herdrchat.core.transcript.MessageSegment
import dev.herdr.herdrchat.ui.theme.HerdrColors
import dev.herdr.herdrchat.ui.theme.formatTime
import dev.herdr.herdrchat.ui.theme.pressScale

/** Coloured presence dot; pulses a soft ring while the agent is working/blocked. */
@Composable
fun PresenceDot(status: AgentStatus, modifier: Modifier = Modifier, ringColor: Color = Color.Transparent) {
    val color = HerdrColors.statusColor(status)
    val pulsing = status == AgentStatus.WORKING || status == AgentStatus.BLOCKED
    Box(modifier.size(14.dp), contentAlignment = Alignment.Center) {
        if (pulsing) {
            val t = rememberInfiniteTransition(label = "pulse")
            val scale by t.animateFloat(
                1f, 2.4f,
                infiniteRepeatable(tween(1400, easing = FastOutSlowInEasing)), label = "s",
            )
            val alpha by t.animateFloat(
                0.45f, 0f,
                infiniteRepeatable(tween(1400, easing = FastOutSlowInEasing)), label = "a",
            )
            Box(
                Modifier
                    .size(10.dp)
                    .graphicsLayer { scaleX = scale; scaleY = scale; this.alpha = alpha }
                    .clip(CircleShape)
                    .background(color),
            )
        }
        Box(
            Modifier
                .size(11.dp)
                .clip(CircleShape)
                .background(ringColor)
                .padding(1.5.dp)
                .clip(CircleShape)
                .background(color),
        )
    }
}

/** Three bouncing dots for a "typing…" indicator. */
@Composable
fun TypingDots(color: Color, dotSize: Dp = 5.dp) {
    val t = rememberInfiniteTransition(label = "typing")
    Row(horizontalArrangement = Arrangement.spacedBy(3.dp), verticalAlignment = Alignment.CenterVertically) {
        repeat(3) { i ->
            val a by t.animateFloat(
                0.25f, 1f,
                infiniteRepeatable(tween(560, delayMillis = i * 140, easing = FastOutSlowInEasing), RepeatMode.Reverse),
                label = "dot$i",
            )
            Box(
                Modifier
                    .size(dotSize)
                    .graphicsLayer { alpha = a }
                    .clip(CircleShape)
                    .background(color),
            )
        }
    }
}

/** A single chat bubble. Outgoing (user) hugs the right; agent bubbles the left.
 *  WhatsApp-style asymmetric corner on the first bubble of a run; the rest of the
 *  run stays fully rounded and tighter-spaced. */
@Composable
fun MessageBubble(message: ChatMessage, groupedWithPrev: Boolean) {
    val dark = isSystemInDarkTheme()
    val outgoing = message.role == ChatMessage.Role.USER
    val round = 17.dp
    val tail = 5.dp
    val shape = if (outgoing) {
        RoundedCornerShape(round, if (groupedWithPrev) round else tail, round, round)
    } else {
        RoundedCornerShape(if (groupedWithPrev) round else tail, round, round, round)
    }
    val bubble = if (outgoing) HerdrColors.outgoingBubble(dark) else HerdrColors.incomingBubble(dark)
    val time = remember(message.id) { formatTime(message.timestamp) }

    Row(
        modifier = Modifier.fillMaxWidth(),
        horizontalArrangement = if (outgoing) Arrangement.End else Arrangement.Start,
    ) {
        Column(
            modifier = Modifier
                .widthIn(max = 328.dp)
                .shadow(1.dp, shape, clip = false)
                .clip(shape)
                .background(bubble)
                .padding(horizontal = 10.dp, vertical = 6.dp),
            verticalArrangement = Arrangement.spacedBy(5.dp),
        ) {
            if (message.agentLabel != null && !outgoing && !groupedWithPrev) {
                Text(
                    text = message.agentLabel,
                    style = MaterialTheme.typography.labelMedium,
                    fontWeight = FontWeight.SemiBold,
                    color = HerdrColors.headerGreen,
                )
            }
            message.segments.forEach { SegmentView(it, dark) }
            if (time != null) {
                Text(
                    text = time,
                    style = MaterialTheme.typography.labelSmall,
                    color = HerdrColors.secondaryText(dark).copy(alpha = 0.8f),
                    modifier = Modifier.align(Alignment.End),
                )
            }
        }
    }
}

@Composable
private fun SegmentView(segment: MessageSegment, dark: Boolean) {
    when (segment) {
        is MessageSegment.Text -> androidx.compose.material3.ProvideTextStyle(MaterialTheme.typography.bodyLarge) {
            MarkdownText(
                markdown = segment.value,
                color = HerdrColors.primaryText(dark),
                onTint = false,
            )
        }
        is MessageSegment.Thinking -> Chip(Icons.Filled.Psychology, "düşündü", HerdrColors.secondaryText(dark))
        is MessageSegment.ToolUse -> Chip(
            Icons.Filled.Build,
            segment.input?.let { "${segment.name}: $it" } ?: segment.name,
            HerdrColors.headerGreen,
        )
        is MessageSegment.ToolResult -> Chip(Icons.Filled.CheckCircle, "araç sonucu", HerdrColors.secondaryText(dark))
    }
}

@Composable
private fun Chip(icon: ImageVector, label: String, tint: Color) {
    Row(
        modifier = Modifier
            .clip(CircleShape)
            .background(tint.copy(alpha = 0.12f))
            .padding(horizontal = 9.dp, vertical = 4.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(5.dp),
    ) {
        Icon(icon, contentDescription = null, tint = tint, modifier = Modifier.size(14.dp))
        Text(
            text = label,
            style = MaterialTheme.typography.labelMedium,
            color = tint,
            maxLines = 1,
            overflow = TextOverflow.Ellipsis,
        )
    }
}

/** A slim, indeterminate "waiting for the reply" bar shown while the agent
 *  works. Transcripts are turn-granular, so there's no finer token stream to
 *  surface — a quiet sweeping bar reads as progress without pretending to show
 *  content it doesn't have (it replaces the old terminal-tail bubble). */
@Composable
fun WaitingBar(modifier: Modifier = Modifier) {
    val t = rememberInfiniteTransition(label = "waiting")
    val phase by t.animateFloat(
        0f, 1f,
        infiniteRepeatable(tween(1100, easing = FastOutSlowInEasing), RepeatMode.Reverse),
        label = "sweep",
    )
    Box(
        modifier
            .fillMaxWidth()
            .height(3.dp)
            .clip(CircleShape)
            .background(HerdrColors.accent.copy(alpha = 0.14f)),
    ) {
        BoxWithConstraints(Modifier.fillMaxHeight()) {
            val segment = maxWidth * 0.28f
            Box(
                Modifier
                    .offset(x = (maxWidth - segment) * phase)
                    .width(segment)
                    .fillMaxHeight()
                    .clip(CircleShape)
                    .background(HerdrColors.accent),
            )
        }
    }
}

/** Quick-reply bar shown when an agent is blocked, waiting for input. */
@Composable
fun BlockedReplyBar(onKeys: (List<String>) -> Unit) {
    val dark = isSystemInDarkTheme()
    val blocked = HerdrColors.statusColor(AgentStatus.BLOCKED)
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .background(blocked.copy(alpha = if (dark) 0.14f else 0.10f))
            .padding(vertical = 10.dp, horizontal = 12.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(6.dp)) {
            TypingDots(color = blocked, dotSize = 4.dp)
            Text(
                text = "Agent yanıt bekliyor",
                style = MaterialTheme.typography.labelLarge,
                fontWeight = FontWeight.SemiBold,
                color = blocked,
            )
        }
        Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
            ReplyButton("Onayla", listOf("Enter"), onKeys)
            ReplyButton("1", listOf("1", "Enter"), onKeys)
            ReplyButton("2", listOf("2", "Enter"), onKeys)
            ReplyButton("Esc", listOf("Escape"), onKeys)
        }
    }
}

@Composable
private fun ReplyButton(title: String, keys: List<String>, onKeys: (List<String>) -> Unit) {
    val dark = isSystemInDarkTheme()
    val interaction = remember { MutableInteractionSource() }
    Box(
        modifier = Modifier
            .pressScale(interaction)
            .shadow(1.dp, CircleShape, clip = false)
            .clip(CircleShape)
            .background(HerdrColors.incomingBubble(dark))
            .clickable(interaction, indication = null) { onKeys(keys) }
            .padding(horizontal = 16.dp, vertical = 9.dp),
    ) {
        Text(
            text = title,
            style = MaterialTheme.typography.bodyMedium,
            fontWeight = FontWeight.Medium,
            color = HerdrColors.primaryText(dark),
        )
    }
}
